### ISSUE TITLE
[ Feature Request ] Encryption key for p1 smart meter

### ISSUE THREAD

**Issue by MSe-5-14** [domoticz/domoticz#2738]
In Luxembourg the smart meter (smarty) is compliant with the "Dutch Smart Meter Requirements v5.0.2 Final P1"  In deviation to DSMR P1, an encryption layer is added over the P1 protocol.
This encryption layer is based on DLMS security suite 0 algorithm: AES128-GCM

Can the decryption be added to domoticz? The encryption keys can be requested by the user from the electricity company. here is some more information that i found:
[https://smarty.creos.net/wp-content/uploads/P1PortSpecification.pdf](https://smarty.creos.net/wp-content/uploads/P1PortSpecification.pdf) 
[https://github.com/mweimerskirch/smarty_dsmr_proxy](https://github.com/mweimerskirch/smarty_dsmr_proxy)
[https://github.com/ndokter/dsmr_parser/issues/27](https://github.com/ndokter/dsmr_parser/issues/27)
[http://www.weigu.lu/microcontroller/smartyreader/index.html](http://www.weigu.lu/microcontroller/smartyreader/index.html)

**Comment by gordonb3** [comment]
Seems there are more differences. Am I correct in understanding that you do not have double tariffs in Luxembourg or am I somehow misinterpreting that P1PortSpecification doc?

I suppose adding a decryptor should not be much of an issue, although there may be some complication with identifying what is a complete message/telegram. The main problem of course is that if someone from outside Luxembourg would want to attack this it will be kind of hard to verify the code before releasing it.

**Comment by MSe-5-14** [comment]
I didn't know there were double tariffs in the dutch specification. (i should maybe read the documentation better :) )
I've requested my decryption keys from my energy supplier. when i receive them, i will then be able to try and decrypt the signal and provide an example of complete message/telegram. If needed i can connect a Raspberry Pi with a VPN connection and give a developer access.

**Comment by gordonb3** [comment]
May also like to provide an example of what the raw message looks like (start-stop tokens?)

Yes, Dutch specification is based on a high (midweek day) tariff and a low (night/weekend) tariff and even though you can get a single tariff contract that will still use the same meter and thus register as either tariff 1 or 2. According to the specification you linked we need to add tariff 0 to the filter list for you to see any data.

**Comment by MSe-5-14** [comment]
I'm still waiting for the decryption keys. I think i might be one of the first people to ask my energy company's helpdesk for the keys. They keep providing me with the p1 port specification document, which i assume they havent read... Ill get back to you when i have more information. i did find another website and project that managed to access the meter [https://www.nexxtlab.lu/smarty-dongle/](https://www.nexxtlab.lu/smarty-dongle/)

**Comment by Netskeh** [comment]
> 
> 
> I'm still waiting for the decryption keys. I think i might be one of the first people to ask my energy company's helpdesk for the keys. 

I feel your pain, which is why, as author of the post you linked, I added the email addresses so people hopefully get their messages to the right person faster. Unfortunately some companies are not yet used to the process, and some are totally unaware of the "P1 key". I waited roughly 4-5 months for mine...

The only piece of information I can give you is to write an email. Talking over phone to their helpdesk will not kick off the required process. Good luck :)

PS: if your energy supplier still does not understand your request, ask your grid operator - they are faster. And usually, as far as I am aware, the energy supplier will have to ask your grid operator anyway...

**Comment by thhatzegopteryx** [comment]
ask Creos by phone for decryption key. They issued my decryption key within 2 working days in June 2018.
Please let me know once you found a way to enter it into domoticz.

**Comment by warpbe** [comment]
Any news on this topic ?

**Comment by MSe-5-14** [comment]
sorry i completely forgot about this. I got my Encryption key and used a pyton script (https://github.com/mweimerskirch/smarty_dsmr_proxy/blob/master/decrypt.py) to decrypt the p1 information. i then use socat to publish the decrypted information on port 2001. I then added P1 Smart Meter with LAN interface in Domoticz. The information is now visible in domoticz. Im just wondering if the information is correct. i somehow doubt that im using 10000 kw/h daily

**Comment by gizmocuz** [comment]
@MSe-5-14 , could it be that you see the values with a multiply of 1000 ?

I could try to implement it, please contact me on the forum with a PM (gizmocuz),
for this I would need the key and a raw capture of multiple packets or access to the P1 port with ser2net

**Comment by MSe-5-14** [comment]
could be...  i've send you a private message on the forum about access to the p1 port. hope to hear from you soon.

**Comment by gizmocuz** [comment]
@MSe-5-14 , send you a email

**Comment by MSe-5-14** [comment]
resolved by @gizmocuz  with commits [#a2d4f914f](https://github.com/domoticz/domoticz/commit/a2d4f914f)  and [#e13d8e451](https://github.com/domoticz/domoticz/commit/e13d8e451). The feature is now available in beta version 12073. Thank you again for the hard work!!

**Comment by gizmocuz** [comment]
Thanks for the help!

**Comment by thhatzegopteryx** [comment]
Do you hae a step by step procedure you could share with me to replicate it on my Lux meter (I already have decryption code).
Thank you.

**Comment by MSe-5-14** [comment]
Really simple. Update your domoticz to version 12073. then go to hardware add the p1 smart meter usb. select your usb port to which you connected the p1 port. set the Baudrate to 115200 and add the decryption key at the bottom. if you dont see the decryption option clear your browser and appdata.(or use private mode on your browser).

**Comment by thhatzegopteryx** [comment]
Thank you. Will try

**Comment by uclaude** [comment]
Hello
Is there any known issue on the decrypt part of the P+.
Since a while, I can't say since which version, it let crash domoticz also with a new installation.
If needed I can provide logs.
Thanks.

**Comment by thhatzegopteryx** [comment]
Hi,

Creos has confirmed that they have performed an update (unclear what type of update) on October 5 which seems to be the root cause.

Have been using Smart meter decryption key on domoticz beta build 12213 from JADAHAL for Synology NAS DS118 since August 10 2020 without any problem until October 5 2020 midnight when domoticz was no longer reading any data neither from Smart meter (USB) nor Pi raspberry (python/json). 

Since then I have tried Domoticz beta build 12213, 12393 and 12494 on Synology.
I noticed that since October 5 2020 domotixz crashes if and only if
  - smart meter is connected via P1 cable and
  - any data is entered into the "decryption key" field
If I unplug the cable and keep a decryption key or if I keep the decryption key blank (or change the baud rate from 115200 to 9600) then domoticz does not crash (and of course no data is read).

We most probably have to fix this issue by changing code in domoticz (as I am noob I have no clue how to do it)

**Comment by thhatzegopteryx** [comment]
I have tried Domoticz beta build 12213, 12393 and 12494 on Synology.
I noticed that since October 5 2020 it crashes if and only if
  - smart meter is connected via P1 cable and
  - any data is entered into the "decryption key" field
If I unplug the cable and keep a decryption key or if I keep the decryption key blank (or change the baud rate from 115200 to 9600) then domoticz does not crash (and of course no data is read).
Could it be that wrong key is causing the issue? This would imply that Creos has changed the decryption key.

**Comment by Netskeh** [comment]
there is a chance that your decryption key changed when they updated your smart meter.
Just send them a new request - however it will always be a new key, no matter if the update changed it or not. They are/ were unable to resend you the same one, instead they generate a new one which they forward to you at the same time.

I suspect domotixz crashing because it is unable to parse the encrypted data?

**Comment by thhatzegopteryx** [comment]
Thank you.
We have opened new conversation #4460 as that one was closed.
Decryption key was not changed. Creos confirmed old key is still valid.
Creos also confirmed that they have made some changes from their side (unclear which one as they did not responded on querry).
gizmocuz is currently working on new script based on input from uclaude.
Thank you to all of you

**Comment by gizmocuz** [comment]
@thhatzegopteryx , I have not started yet... but will try today. As I do not have a valid decryption key I wonder what will happen

**Comment by gizmocuz** [comment]
I think I found the problem. We (I) was assuming a maximum complete packet length of 1000 bytes, but the first (new) packet I tested was 1019, this resulted in invalid memory readings in the underlying SSL library and caused a crash.
Increased this value to 2048 now... way enough for feature packets...
There is a new beta (12689) that should solve the problem. It is already available for Arm systems (raspberry) , for Windows/Debian it will be ready within the hour.
Is it possible to test ?
